### DISCUSSION TITLE
Dokku - Django deployment issue

### DISCUSSION THREAD

**Discussion by amirexe** [dokku/dokku#5334] [Q&A] [answered]
Hi there,
Thank you for the great efforts on Dokku :)

For my question, I have a new dedicated server (Linux Centos 7), installed Dokku and all related configurations, tried a static website deployment and it’s working fine.

However, I have tried to deploy a Django application (btw I am using Dokku 0.26.6), with all related Django configurations and Python version python-3.9.9 (also tried python-3.8.13, python-3.9.13) in the runtime.txt file, but when tried to push, I get the failure message always about Python version cannot be found in the stack (which is heroku-20)”

and the push is always stucked in the step of updating the python runtime for around 4 minutes, then the failure message

Please note that the same installation is working fine on another old server since Dokku 0.26.6 was released.

Appreciate your support.

Many thanks

**Comment by amirexe** [comment]
BUILD_ENV added successfully
-----> Python app detected
-----> Using Python version specified in runtime.txt
       !     Requested runtime (python-3.9.13) is not available for this stack (heroku-20).
       !     Aborting.  More info: https://devcenter.heroku.com/articles/python-support

**Comment by josegonzalez** [comment, marked as answer]
If the runtime installation is stalling and eventually failing, that sounds like maybe there is a networking issue between your server and S3.

Mind upgrading herokuish? I don't know what the commands are for that on CentOS (and we no longer support that as of 0.28.x) but updating that package might help.

**Reply by amirexe** [reply]
Thank you Jose.
Regarding the network issue, is there any command that can help to identify if there is a network issue or not?

**Reply by josegonzalez** [reply]
You could turn on buildpack debugging via:

```shell
# where $APP is your app name
dokku config:set $APP BUILDPACK_XTRACE=1
```

And trigger a deploy. It should show a curl command at some point that is stalling, which you can then execute on your own on that server to see whats going on.

**Reply by amirexe** [reply]
We have tried that and got the following curl command which the push is stalling at:

curl --output /dev/null --silent --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz

We have tried that curl command exactly on the server directly and worked with no issue.
Appreciate your advice.

**Reply by josegonzalez** [reply]
Mind running the following commands and posting the output?

```shell
docker run --rm -it gliderlabs/herokuish:latest curl --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz

docker run --rm -it ubuntu:latest curl --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz
```

**Reply by amirexe** [reply]
docker run --rm -it gliderlabs/herokuish:latest curl --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz

Warning: Transient problem: timeout Will retry in 1 seconds. 3 retries left.
Warning: Transient problem: timeout Will retry in 2 seconds. 2 retries left.
Warning: Transient problem: timeout Will retry in 4 seconds. 1 retries left.
curl: (28) Resolving timed out after 5514 milliseconds

**Reply by amirexe** [reply]
docker run --rm -it ubuntu:latest curl --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz

Unable to find image 'ubuntu:latest' locally
latest: Pulling from library/ubuntu
d19f32bd9e41: Pull complete 
Digest: sha256:34fea4f31bf187bc915536831fd0afc9d214755bf700b5cdb1336c82516d154e
Status: Downloaded newer image for ubuntu:latest
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "curl": executable file not found in $PATH: unknown.

**Reply by josegonzalez** [reply]
Ah try:

```shell
docker run --rm -it curlimages/curl curl --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz
```

**Reply by amirexe** [reply]
docker run --rm -it curlimages/curl curl --head --fail --retry 3 --retry-connrefused --connect-timeout 5 https://heroku-buildpack-python.s3.us-east-1.amazonaws.com/heroku-20/runtimes/python-3.9.9.tar.gz

Unable to find image 'curlimages/curl:latest' locally
latest: Pulling from curlimages/curl
df9b9388f04a: Pull complete 
03e93fb85e57: Pull complete 
236a88e52ee2: Pull complete 
cd50f4fd3a4e: Pull complete 
9d1b8d95f293: Pull complete 
2b56d1b81134: Pull complete 
2bb5a2891534: Pull complete 
22ceed0c0fbf: Pull complete 
dfa9a3fac73e: Pull complete 
f78b99329c8f: Pull complete 
a32f5307f213: Pull complete 
Digest: sha256:5a2a25d96aa941ea2fc47acc50122f7c3d007399a075df61a82d6d2c3a567a2b
Status: Downloaded newer image for curlimages/curl:latest
curl: (28) Resolving timed out after 5000 milliseconds
Warning: Problem : timeout. Will retry in 1 seconds. 3 retries left.
curl: (28) Resolving timed out after 5000 milliseconds
Warning: Problem : timeout. Will retry in 2 seconds. 2 retries left.
curl: (28) Resolving timed out after 5000 milliseconds
Warning: Problem : timeout. Will retry in 4 seconds. 1 retries left.
curl: (28) Resolving timed out after 5000 milliseconds

**Reply by josegonzalez** [reply]
Okay it seems like there is some problem resolving that domain from inside of a container. What hosting provider are we using?

**Reply by amirexe** [reply]
It's hosted on Linode.

However,
- We tried that curl command on the server directly with no issues.
- We have another Linode server with the same installation which is working fine since Dokku 0.26.6 was released.

**Reply by amirexe** [reply]
Hi Jose,
Thanks for your support, it helped much.
Please note that the issue seems to be resolved by the following commands on the server

sudo pkill docker
sudo service docker restart

**Reply by josegonzalez** [reply]
Nice yeah, that was gonna be my guess (docker was wedged somehow).

You can also enable live-restore so you don't lose containers during a docker restart: https://docs.docker.com/config/containers/live-restore/

**Reply by amirexe** [reply]
Great, thank you so much :)
